### ISSUE TITLE
does not support changing ip addresses

### ISSUE THREAD

**Issue by zetaab** [deforestg/django-cluster-redis#2]
Hi

I tried this library in top of kubernetes and it works pretty well. However, the problem starts when I restart containers in redis cluster. Kubernetes default behaviour is that the pod is changing ip address always when the pod starts.

So what I did: I restarted whole cluster
```
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/site-packages/django/core/handlers/exception.py", line 41, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.5/site-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/usr/local/lib/python3.5/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/code/core/views.py", line 12, in home
    tmp = cache.get(keyname)
  File "/usr/local/lib/python3.5/site-packages/django_redis/cache.py", line 40, in _decorator
    raise e.parent
  File "/usr/local/lib/python3.5/site-packages/django_redis/client/default.py", line 201, in get
    value = client.get(key)
  File "/usr/local/lib/python3.5/site-packages/django_cluster_redis/cache.py", line 43, in get
    return self._action('get', *args, **kwargs)
  File "/usr/local/lib/python3.5/site-packages/django_cluster_redis/cache.py", line 9, in _action
    return method(*args, **kwargs)
  File "/usr/local/lib/python3.5/site-packages/redis/client.py", line 976, in get
    return self.execute_command('GET', name)
  File "/usr/local/lib/python3.5/site-packages/redis/client.py", line 673, in execute_command
    connection.send_command(*args)
  File "/usr/local/lib/python3.5/site-packages/redis/connection.py", line 610, in send_command
    self.send_packed_command(self.pack_command(*args))
  File "/usr/local/lib/python3.5/site-packages/redis/connection.py", line 585, in send_packed_command
    self.connect()
  File "/usr/local/lib/python3.5/site-packages/redis/connection.py", line 489, in connect
    raise ConnectionError(self._error_message(e))
redis.exceptions.ConnectionError: Error 113 connecting to 10.1.24.153:6379. Host is unreachable.
```

So what happens here? There is some cache which remembers the old address and tries to connect to it. However, it should delete addresses from cache if those are not working. After not working ones are deleted it should load new addresses from original source which in my case is:

```
CACHES = {
    "default": {
        "BACKEND": "django_redis.cache.RedisCache",
        "LOCATION": "redis://redis-cluster:6379/0",
        "OPTIONS": {
            'REDIS_CLIENT_CLASS': 'django_cluster_redis.cache.ClusterRedis',
        }
    }
}
```

**Comment by deforestg** [comment]
So the caching that remembers the connection is purely in memory, so restarting the process should reset that. Are you saying that you want to be able to continue to run the same process and delete connections from that pool when they fail?

**Comment by jordanpwalsh** [comment]
@deforestg I believe he's saying that the pod addresses for, I'm assuming redis, will change when the pods are restarted and/or changed. This would be expected behavior. I have a feeling only Redis was restarted here. 

@zetaab what pods were restarted to cause this? Just Redis, others, etc.?

**Comment by zetaab** [comment]
I mean that when restarting redis pods the redis ip addresses is changed.
After that application still uses old ips and connection does not work at
all.

16.11.2017 4.32 ap. "Jordan Walsh" <notifications@github.com> kirjoitti:

**Comment by deforestg** [comment]
So I'm trying to understand what behavior you're expecting it to have. If you restart your redis cluster and the IPs change, is your app notified of this? Do you want this to delete failed connections as they happen and automatically retry?

**Comment by zetaab** [comment]
when i configure "LOCATION": "redis://redis-cluster:6379/0", to application - I do except that it should use that location to (re)fetch information from locations if there is problems connecting to redis. App is not notified of this, but app should IMO do something if it sees that "this server is not working anymore". There is no idea to keep connections cached if those does not work anymore? If I can choose, I would remove cached connections if those does not work anymore. After cached connections is removed, the app should refetch information from configured original location in config?

**Comment by jordanpwalsh** [comment]
If I may opine:

 I'm highly suspecting some kind of cached DNS and/or other DNS related issue in your kubernetes cluster. How are you running it? Rancher, OpenShift, rolling your own, etc. The only thing this library does is connect to what you've told it to connect to. Removing containers/orchestration from the conversation and replacing them with traditional servers/vms would sound like this: "I deleted and  recreated my redis cluster running on 3 vm/servers, and now I'm unable to connect to the new cluster because the IP addresses resolve to unavailable hosts". I would tell you to look at DNS, and kubernetes is the system that manages that. 

Put simply: if you add a host-name in your LOCATIONS array, it's going to query the system to get those IPs. It's my opinion that django-redis-cluster, nor its upstream dependencies, should interfere with how an IP address is resolved from DNS.

**Comment by deforestg** [comment]
Also simply restarting whatever wsgi you're running would reset all the connections in the pool
